internal/deployment_client.go

```diff
@@ -0,0 +1,244 @@
+		LastUpdateTime time.Time
+	}
+
+	// DeploymentMetadataUpdate modifies user-defined metadata entries that describe
+	// a deployment.
+	// NOTE: Experimental
+	DeploymentMetadataUpdate struct {
+		// UpsertEntries - Metadata entries inserted or modified. When values are not
+		// of type *commonpb.Payload, the client data converter will be used to generate
+		// payloads.
+		UpsertEntries map[string]interface{}
+
+		// RemoveEntries - List of keys to remove from the metadata.
+		RemoveEntries []string
+	}
+
+	// DeploymentSetCurrentOptions provides options for [DeploymentClient.SetCurrent].
+	// NOTE: Experimental
+	DeploymentSetCurrentOptions struct {
+		// Deployment - An identifier for this deployment.
+		Deployment Deployment
+
+		// MetadataUpdate - Optional: Changes to the user-defined metadata entries
+		// for this deployment.
+		MetadataUpdate DeploymentMetadataUpdate
+	}
+
+	// DeploymentSetCurrentResponse is the response type for [DeploymentClient.SetCurrent].
+	// NOTE: Experimental
+	DeploymentSetCurrentResponse struct {
+		// Current - Information about the current deployment after this operation.
+		Current DeploymentInfo
+
+		// Previous - Information about the last current deployment, i.e., before this operation.
+		Previous DeploymentInfo
+	}
+
+	// DeploymentDescribeOptions provides options for [DeploymentClient.Describe].
+	// NOTE: Experimental
+	DeploymentDescribeOptions struct {
+		// Deployment - Identifier that combines the deployment series name with their Build ID.
+		Deployment Deployment
+	}
+
+	// DeploymentGetReachabilityOptions provides options for [DeploymentClient.GetReachability].
+	// NOTE: Experimental
+	DeploymentGetReachabilityOptions struct {
+		// Deployment - Identifier that combines the deployment series name with their Build ID.
+		Deployment Deployment
+	}
+
+	// DeploymentGetCurrentOptions provides options for [DeploymentClient.GetCurrent].
+	// NOTE: Experimental
+	DeploymentGetCurrentOptions struct {
+		// SeriesName - Name of the deployment series.
+		SeriesName string
+	}
+
+	// DeploymentClient is the client that manages deployments.
+	// NOTE: Experimental
+	DeploymentClient interface {
+		// Describes an existing deployment.
+		// NOTE: Experimental
+		Describe(ctx context.Context, options DeploymentDescribeOptions) (DeploymentInfo, error)
+
+		// List returns an iterator to enumerate deployments in the client's namespace.
+		// It can also optionally filter deployments by series name.
+		// NOTE: Experimental
+		List(ctx context.Context, options DeploymentListOptions) (DeploymentListIterator, error)
+
+		// GetReachability returns reachability information for a deployment. This operation is
+		// expensive, and results may be cached. Use the returned [DeploymentReachabilityInfo.LastUpdateTime]
+		// to estimate cache staleness.
+		// NOTE: Experimental
+		GetReachability(ctx context.Context, options DeploymentGetReachabilityOptions) (DeploymentReachabilityInfo, error)
```

Comment:
Why would I ever call `Describe` when this has everything describe does and more? Why even have `Describe` or better yet, why not make `Describe` return reachability information?

Reply:
Getting Reachability info is a very expensive operation, so it may be stale and/or rate limited, and that's why they provided a "light" version with `Describe` that only returns the easy bits... I will add a comment.

Reply:
Hrmm, to me it sounds like overlapping concerns and to obtain reachability should just be an option on describe. I assume this was discussed at the API level and decided to separate `DescribeDeployment` and `GetDeploymentReachability` instead of just make this an option on `Describe` (like we have on `DescribeTaskQueue`)? If it was discussed, can you share the reasoning there?

Reply:
Yes, it was discussed, they can only do rate limiting at the method level, and that's why they want a separate method.

Reply:
Hrmm, that is unfortunate to burden a user API all the way out through the SDK with our server limitations, but ok. I am guessing they decided it was ok not to need to separately rate limit the different `DescribeTaskQueueRequest` options.